### ISSUE TITLE
Deserialization of data formerly using `PersistDbSpecific` and now using `PersistLiteralEscaped`

### ISSUE THREAD

**Issue by friedbrice** [yesodweb/persistent#1161]
Persistent 2.11 deprecated `PersistDbSpecific` due to inconsistent handling of escaping among SQL backends (which went so far as to even cause the examples provided in the documentation to fail), replacing it with `PersistLiteral` and `PersistLiteralEscaped`. Users are directed to replace occurrences of `PersistDbSpecific` with a constructor appropriate for their intended purpose. The advice is misleading, however, because doing so will cause data created with prior versions of persistent to fail to deserialize in some cases.

I have constructed [an example](https://github.com/friedbrice/persist_ser). While it's a bit complicated, it does demonstrate data that successfully makes a round trip in Persistent 2.10 but that fails to deserialize once Persistent has been upgraded to 2.11. My attempts to find a simpler example have so far failed. Users can sidestep the problem if they are willing to keep `PersistDbSpecific` in their code and put up with the deprecation warnings, but this confounds our attempts to migrate users away from it. The alternative is to ask users to do a costly and manual database migration, which is less than ideal for its own reasons.

The de/serialization in question is defined by the `FromJSON`/`ToJSON` instances for `PersistValue` in [Database.Persist.Types.Base](https://github.com/yesodweb/persistent/blob/master/persistent/Database/Persist/Types/Base.hs). I believe that we can provide a backwards-compatible format, allowing users to migrate away from `PersistDbSpecific` without needing a costly migration, by liberalizing the deserializer so that `p`-prefixed byte arrays are accepted as valid `PersistLiteral` and `PersistLiteralEscaped` values, though the details are necessarily backend specific, as Postgres will need to deserialize such values to `PersistLiteralEscaped` while MySQL and Sqlite will need to deserialize such values to `PersistLiteral`. If this sounds like an acceptable approach to the maintainers, I can begin work on a patch this weekend.

**Comment by friedbrice** [comment]
^ @MaxGabriel and @parsonsmatt

**Comment by parsonsmatt** [comment]
Sounds good to me! Thanks for finding and proposing a fix for this.

**Comment by parsonsmatt** [comment]
[Potential fix for postgres, but potential break for mysql and sqlite](https://github.com/friedbrice/persistent/commit/f2fb5475d26487c0ec017004149fcd11a7e742cc).

As mentioned, the problem is due to `ToJSON` and `FromJSON` instances for `PersistValue`. These are 

The example code follows the deprecation warning and replaces with the appropriate constructor in *both* serialization and deserialization. It specifically uses a `[X]` field. `persistent` uses the `PersistList` constructor for lists, and converts the contents into their constitution `PersistValue`s.

```haskell
instance {-# OVERLAPPABLE #-} PersistField a => PersistField [a] where
    toPersistValue = PersistList . fmap toPersistValue
    fromPersistValue (PersistList l) = fromPersistList l
    fromPersistValue (PersistText t) = fromPersistValue (PersistByteString $ TE.encodeUtf8 t)
    fromPersistValue (PersistByteString bs)
        | Just values <- A.decode' (L.fromChunks [bs]) = fromPersistList values
    -- avoid the need for a migration to fill in empty lists.
    -- also useful when Persistent is not the only one filling in the data
    fromPersistValue (PersistNull) = Right []
    fromPersistValue x = Left $ fromPersistValueError "List" "list, string, bytestring or null" x
```

`PersistList` is serialized and deserialized by the client packages. And the `ToJSON`-based encoding is done by `listToJSON` (and `mapToJSON`), both of which delegate to `toJsonText`. So we can track uses of these functions to determine what the issue is.

Grepping for `listToJSON`, we get several results:

```haskell
-- Postgresql
instance PGTF.ToField P where
    toField (P (PersistText t))        = PGTF.toField t
    toField (P (PersistByteString bs)) = PGTF.toField (PG.Binary bs)
    toField (P (PersistInt64 i))       = PGTF.toField i
    toField (P (PersistDouble d))      = PGTF.toField d
    toField (P (PersistRational r))    = PGTF.Plain $
                                         BBB.fromString $
                                         show (fromRational r :: Pico) --  FIXME: Too Ambigous, can not select precision without information about field
    toField (P (PersistBool b))        = PGTF.toField b
    toField (P (PersistDay d))         = PGTF.toField d
    toField (P (PersistTimeOfDay t))   = PGTF.toField t
    toField (P (PersistUTCTime t))     = PGTF.toField t
    toField (P PersistNull)            = PGTF.toField PG.Null
    toField (P (PersistList l))        = PGTF.toField $ listToJSON l
    toField (P (PersistMap m))         = PGTF.toField $ mapToJSON m
    toField (P (PersistDbSpecific s))  = PGTF.toField (Unknown s)
    toField (P (PersistLiteral l))     = PGTF.toField (UnknownLiteral l)
    toField (P (PersistLiteralEscaped e)) = PGTF.toField (Unknown e)
    toField (P (PersistArray a))       = PGTF.toField $ PG.PGArray $ P <$> a
    toField (P (PersistObjectId _))    =
        error "Refusing to serialize a PersistObjectId to a PostgreSQL value"

-- MySQL
instance MySQL.Param P where
    render (P (PersistText t))        = MySQL.render t
    render (P (PersistByteString bs)) = MySQL.render bs
    render (P (PersistInt64 i))       = MySQL.render i
    render (P (PersistDouble d))      = MySQL.render d
    render (P (PersistBool b))        = MySQL.render b
    render (P (PersistDay d))         = MySQL.render d
    render (P (PersistTimeOfDay t))   = MySQL.render t
    render (P (PersistUTCTime t))     = MySQL.render t
    render (P PersistNull)            = MySQL.render MySQL.Null
    render (P (PersistList l))        = MySQL.render $ listToJSON l
    render (P (PersistMap m))         = MySQL.render $ mapToJSON m
    render (P (PersistRational r))    =
      MySQL.Plain $ BBB.fromString $ show (fromRational r :: Pico)
      -- FIXME: Too Ambiguous, can not select precision without information about field
    render (P (PersistDbSpecific s))    = MySQL.Plain $ BBS.fromByteString s
    render (P (PersistLiteral l))     = MySQL.Plain $ BBS.fromByteString l
    render (P (PersistLiteralEscaped e)) = MySQL.Escape e
    render (P (PersistArray a))       = MySQL.render (P (PersistList a))
    render (P (PersistObjectId _))    =
        error "Refusing to serialize a PersistObjectId to a MySQL value"

-- SQLite
bind :: Statement -> [PersistValue] -> IO ()
bind statement sqlData = do
  mapM_ (\(parameterIndex, datum) -> do
          case datum of
            PersistInt64 int64 -> bindInt64 statement parameterIndex int64
            PersistDouble double -> bindDouble statement parameterIndex double
            PersistRational rational -> bindText statement parameterIndex $ pack $ show (fromRational rational :: Pico)
            PersistBool b -> bindInt64 statement parameterIndex $
                                if b then 1 else 0
            PersistText text -> bindText statement parameterIndex text
            PersistByteString blob -> bindBlob statement parameterIndex blob
            PersistNull -> bindNull statement parameterIndex
            PersistDay d -> bindText statement parameterIndex $ pack $ show d
            PersistTimeOfDay d -> bindText statement parameterIndex $ pack $ show d
            PersistUTCTime d -> bindText statement parameterIndex $ pack $ format8601 d
            PersistList l -> bindText statement parameterIndex $ listToJSON l
            PersistMap m -> bindText statement parameterIndex $ mapToJSON m
            PersistDbSpecific s -> bindText statement parameterIndex $ decodeUtf8With lenientDecode s
            PersistArray a -> bindText statement parameterIndex $ listToJSON a -- copy of PersistList's definition
            PersistObjectId _ -> P.error "Refusing to serialize a PersistObjectId to a SQLite value"

            -- I know one of these is broken, but the docs for `sqlite3_bind_text` aren't very illuminating.
            PersistLiteral l -> bindText statement parameterIndex $ decodeUtf8With lenientDecode l
            PersistLiteralEscaped e -> bindText statement parameterIndex $ decodeUtf8With lenientDecode e
            )
       $ zip [1..] sqlData
  return ()
```

So, there are three levels of codecs:

1. `PersistField` instance: `a -> PersistValue`
2. `ToJSON` instance: `PersistValue -> Value`
3. Database-specific: `Value -> DatabaseType`

We have hardcoded that the `PersistValue -> Value` codec prepend a single character tag to the stringy constructors of `PersistValue`. So we succeed - easily - going from `DatabaseType -> Value -> PersistValue`. The problem is that we fail at the `PersistField` level, and it's not trivial to make that succeed easily. 

So, a previous version of `persistent` stores the JSON with a `PersistDbSpecific` as `"pfoobar"`. The user upgrades `persistent`, and follows the deprecation warnings, and now they're matching on `PersistLiteralEscaped`. The `FromJSON` instance sees `"pfoobar"`, turns it into `PersistDbSpecific "foobar"`, and passes it up to `fromPersistValue`, which probably errors since the `PersistDbSpecific` case was turned into `PersistLiteralEscaped`. Now you get a deserialization error.

As noted, end users can work around this by accepting a `PersistDbSpecific` case for these types, and returning the appropriate value. To patch the example, we'd do:

```diff
 instance PersistField X where
   toPersistValue (X x) = PersistLiteralEscaped $ TE.encodeUtf8 x
   fromPersistValue (PersistLiteralEscaped bs) = Right . X $ TE.decodeUtf8 bs
+  fromPersistValue (PersistDbSpecific bs) = Right . X $ TE.decodeUtf8 bs
   fromPersistValue x = Left $ "Failed to deserialize " <> tshow x
```

Adding that case will allow the `PersistField` instance to properly parse the old data format. This gives us our first todo:

- [ ] Amend the deprecation warning to mention this issue!

Next, we need to reliably be able to determine if there is anything in the schema that runs into this problem. Using `[Type]` in a schema file is a sure bet (unless it's wrapped in another type, like `xs (Jsonb [X])`, *and* that the `newtype` does something other than call `toJsonText` on it). Likewise, using `Map` in a schema directly will do it, too. Regrettably, embedding an entity directly into another one will *also* do it.

```haskell
entityToPersistValueHelper :: (PersistEntity record) => record -> PersistValue
entityToPersistValueHelper entity = PersistMap $ zip columnNames fieldsAsPersistValues
    where
        columnNames = map (unFieldNameHS . fieldHaskell) (entityFields (entityDef (Just entity)))
        fieldsAsPersistValues = map toPersistValue $ toPersistFields entity
```

Unfortunately, this is going to affect anything that encodes a `PersistList` or `PersistMap` with any values containing a `PersistDbSpecific`. I think we'd need to do something like:

1. Pull the `toPersistValue` definition and see if it refers to `PersistDbSpecific`, `PersistList`, or `PersistMap`
    a. If `PersistDbSpecific`, then we need to ensure that type is not stored in a `PersistList` or `PersistMap` anywhere in the schema.
    b. If `PersistList` or `PersistMap`, then we need to ensure that none of the contained values have a `PersistDbSpecific` in them

🤔 

We can't actually do this. While Template Haskell nominally supports [pulling the RHS from a value declaration](https://hackage.haskell.org/package/template-haskell-2.16.0.0/docs/Language-Haskell-TH.html#t:Info), it has not been implemented due to "lack of interest." [Maybe worthwhile to at least modify the comment, since seven use cases have been posted to the GitLab issue](https://gitlab.haskell.org/ghc/ghc/-/issues/1831).

🤔 

At Mercury, a task was developed that would read every row out of the database to verify that the `persistent` schema (as understood by the Haskell code) would work for all the fields and values currently in the database. We could generalize this - given the `entityList :: [EntityDef]`, load each row from each table and verify that parsing succeeds.

If parsing fails, then we can report that to the end user, and they can do *something* with it. Maybe we can detect a `PersistDbSpecific` -> `PersistLiteral` failure and replace the `b` with an `l`. 

But this change would require that the codebase adopt the "breaking" version of the `PersistField` instance - if they have the backwards compatible instance in place, then it'll Work.

Going to continue brainstorming here...

**Comment by parsonsmatt** [comment]
Another possibility is to alter `PersistLiteral` to accept another field, `Espaced | Unescaped | DbSpecific`. The three existing constructors would become pattern synonyms, and parsing/deserializing code would just always succeed. Ex:

```haskell
data PersistField
    = PersistLiteral LiteralType ByteString

data LiteralType = Unescaped | Escaped | DbSpecific

pattern PersistDbSpecific bs <- PersistLiteral _ bs where
  PersistDbSpecific bs = PersistLiteral DbSpecific bs
  
-- etc
```

This would allow an instance like this to round-trip:

```haskell
instance PersistField X where
  toPersistValue (X x) = PersistLiteral Escaped x
  fromPersistValue (PersistDbSpecific x) = X x
```

... which is essentially what we're asking from the datbase layer.

Encoding the values would case on the `Escaped/Unescaped/DbSpecific`, of course. And the constructor for `DbSpecific` could be deprecated, pointing to the `Escaped/Unescaped` constructors. 🤔 

Allowing the patterns to succeed regardless of the contents of the tag would be a bit tricky. If there's code in the wild like:

```haskell
fromField pv = case pv of
    PersistDbSpecific a -> doThing a
    PersistLiteralEscaped b -> doOtherThing b
    PersistLiteral _ c -> doYetAnotherThing c
```

then that would always match on `PersistDbSpecific`, whereas now it would be able to work on the other constructors and do different things with it.

We can make it *not matter* how a value is deserialized from `PersistValue` into a thing, which fixes the problem, but then any behavior relying on *which* `Literal` constructor is used will break.

**Comment by parsonsmatt** [comment]
Ah, a problem emerges:

```haskell
bind :: Statement -> [PersistValue] -> IO ()
bind statement sqlData = do
  mapM_ (\(parameterIndex, datum) -> do
          case datum of
            PersistInt64 int64 -> bindInt64 statement parameterIndex int64
            PersistDouble double -> bindDouble statement parameterIndex double
            PersistRational rational -> bindText statement parameterIndex $ pack $ show (fromRational rational :: Pico)
            PersistBool b -> bindInt64 statement parameterIndex $
                                if b then 1 else 0
            PersistText text -> bindText statement parameterIndex text
            PersistByteString blob -> bindBlob statement parameterIndex blob
            PersistNull -> bindNull statement parameterIndex
            PersistDay d -> bindText statement parameterIndex $ pack $ show d
            PersistTimeOfDay d -> bindText statement parameterIndex $ pack $ show d
            PersistUTCTime d -> bindText statement parameterIndex $ pack $ format8601 d
            PersistList l -> bindText statement parameterIndex $ listToJSON l
            PersistMap m -> bindText statement parameterIndex $ mapToJSON m
            PersistDbSpecific s -> bindText statement parameterIndex $ decodeUtf8With lenientDecode s
            PersistArray a -> bindText statement parameterIndex $ listToJSON a -- copy of PersistList's definition
            PersistObjectId _ -> P.error "Refusing to serialize a PersistObjectId to a SQLite value"

            -- I know one of these is broken, but the docs for `sqlite3_bind_text` aren't very illuminating.
            PersistLiteral l -> bindText statement parameterIndex $ decodeUtf8With lenientDecode l
            PersistLiteralEscaped e -> bindText statement parameterIndex $ decodeUtf8With lenientDecode e
            )
       $ zip [1..] sqlData
  return ()
```

This could would always hit `PersistDbSpecific` and never fall down to `PersistLiteral` or `PersistLiteralEscaped`. Since the code is internal to `persistent`, we can just fix that by casing on `LiteralType`, but it's definitely a demonstration on how this 'fix' can go wrong.

**Comment by parsonsmatt** [comment]
Another gotcha:

```haskell
instance PersistField PgInterval where
    toPersistValue = PersistLiteralEscaped . pgIntervalToBs
    fromPersistValue (PersistDbSpecific bs) = fromPersistValue (PersistLiteralEscaped bs)
    fromPersistValue x@(PersistLiteralEscaped bs) =
      case P.parseOnly (P.signed P.rational <* P.char 's' <* P.endOfInput) bs of
        Left _  -> Left $ fromPersistValueError "PgInterval" "Interval" x
        Right i -> Right $ PgInterval i
    fromPersistValue x = Left $ fromPersistValueError "PgInterval" "Interval" x
```

This instance loops - because `PersistDbSpecific` is going to catch when used with `PersistLiteralEscaped`.

**Comment by parsonsmatt** [comment]
This was fixed by #1205
